internal/rbd/controllerserver.go

```diff
@@ -786,38 +825,9 @@ func (cs *ControllerServer) DeleteVolume(
 	rbdVol, err := genVolFromVolID(ctx, volumeID, cr, req.GetSecrets())
 	defer rbdVol.Destroy()
 	if err != nil {
-		if errors.Is(err, util.ErrPoolNotFound) {
-			log.WarningLog(ctx, "failed to get backend volume for %s: %v", volumeID, err)
-
-			return &csi.DeleteVolumeResponse{}, nil
-		}
-
-		// if error is ErrKeyNotFound, then a previous attempt at deletion was complete
-		// or partially complete (image and imageOMap are garbage collected already), hence return
-		// success as deletion is complete
-		if errors.Is(err, util.ErrKeyNotFound) {
-			log.WarningLog(ctx, "Failed to volume options for %s: %v", volumeID, err)
-
-			return &csi.DeleteVolumeResponse{}, nil
-		}
-
-		// All errors other than ErrImageNotFound should return an error back to the caller
-		if !errors.Is(err, ErrImageNotFound) {
-			return nil, status.Error(codes.Internal, err.Error())
-		}
-
-		// If error is ErrImageNotFound then we failed to find the image, but found the imageOMap
-		// to lead us to the image, hence the imageOMap needs to be garbage collected, by calling
-		// unreserve for the same
-		if acquired := cs.VolumeLocks.TryAcquire(rbdVol.RequestName); !acquired {
-			log.ErrorLog(ctx, util.VolumeOperationAlreadyExistsFmt, rbdVol.RequestName)
-
-			return nil, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, rbdVol.RequestName)
-		}
-		defer cs.VolumeLocks.Release(rbdVol.RequestName)
-
-		if err = undoVolReservation(ctx, rbdVol, cr); err != nil {
-			return nil, status.Error(codes.Internal, err.Error())
+		delRes, err2, done := cs.checkErrAndUndoReserve(ctx, err, volumeID, rbdVol, cr)
```

Comment:
do we need 3 variables here cant we have 2 variables for DeleteVolumeResponse and err?

Reply:
@Madhu-1 then we need few more condition checks below for values of DeleteVolumeResponse and err as there are some combinations from the callee. this make it clean.

Reply:
cant we just return the response we got from the `checkErrAndUndoReserve` functions directly?

Reply:
reduced to just 2 the response and error.. ptal.